File: ruby/test/string_test.rb
require 'minitest/autorun'

describe "string" do

  it "compose string" do
    assert_equal "apple orange", "#{"apple"} #{"orange"}"
  end

  it "multiline string" do
    text = <<-EOS
apple
orange
    EOS
    assert_equal "apple\norange\n", text
  end

  it "int to string" do
    assert_equal "1", 1.to_s
  end

  it "regex match" do
    # http://ruby-doc.org/core-2.2.0/Regexp.html
    assert_equal 0, "apple orange " =~ /apple/
    assert_equal 6, "apple orange " =~ /orange/
    assert_equal nil, "apple orange " =~ /kiwi/
  end

  it "regex matches" do
    assert_equal(
      ["apple orange", "apple", "orange"],
      "apple orange ".match(/(apple).*(orange)/).to_a
    )
  end

  it "regex replace" do
    assert_equal "orange apple ", "apple orange ".gsub(/(apple).*(orange)/, '\2 \1')
  end

  it "concat" do
    text = text_backup = "apple"
    text += " orange "
    assert_equal "apple orange ", text
    assert text != text_backup
  end

  it "concat mutate" do
    text = text_backup = "apple"
    text << " orange "
    assert_equal "apple orange ", text
    assert text.equal?(text_backup)
  end

  it "length" do
    assert_equal 13, "apple orange ".length
  end

  it "format" do
    # http://ruby-doc.org/core-2.2.0/Kernel.html#format-method
    assert_equal "10.0", "%.1f" % 10
    # http://ruby-doc.org/core-2.2.0/Time.html#strftime-method
    assert_equal "2000-01-01", Time.new(2000,1,1).strftime("%Y-%m-%d")
  end

  it "split" do
    assert_equal ["apple", "orange"], "apple orange ".split
    assert_equal ["apple", "orange "], "apple,orange ".split(",")
  end

  it "strip" do
    assert_equal "a b", " a b ".strip
  end

  it "eql?" do
    assert_equal true, "apple" == "apple"
  end

end

File: elixir/test/string_test.exs
ExUnit.start

defmodule StringTest do
  use ExUnit.Case, async: true

  test "compose string" do
    assert "apple orange" = "#{"apple"} #{"orange"}"
  end

  test "multiline string" do
    text = """
apple
orange
"""
    assert "apple\norange\n" = text
  end

  test "int to string" do
    assert "1", Integer.to_string(1)
  end

  test "regex match" do
    # http://elixir-lang.org/docs/v1.0/elixir/Regex.html
    assert [{0, _}] = Regex.run(~r/apple/, "apple orange ", return: :index)
    assert [{6, _}] = Regex.run(~r/orange/, "apple orange ", return: :index)
    assert nil == Regex.run(~r/kiwi/, "apple orange ", return: :index)
  end

  test "regex matches" do
    assert ["apple orange", "apple", "orange"] =
      Regex.run(~r/(apple).*(orange)/, "apple orange ")
  end

  test "regex replace" do
    assert "orange apple " =
      Regex.replace(~r/(apple).*(orange)/, "apple orange ", "\\2 \\1")
  end

  test "concat" do
    assert "apple orange " = "apple" <> " orange "
  end

  test "concat mutate" do
    # no mutation
  end

  test "length" do
    assert 13 = String.length("apple orange ")
  end

  test "format" do
    # TODO
  end

  test "split" do
    assert ["apple", "orange"] = String.split("apple orange ")
    assert ["apple", "orange "] = String.split("apple,orange ", ",")
  end

  test "strip" do
    assert "a b" = String.strip(" a b ")
  end

  test "eql?" do
    assert "apple" = "apple"
  end

end
